### PR TITLE
Component docs refactor

### PR DESCRIPTION
**Pull request by benjgil**
## Proposed Changes

- split out the various core react docs template sections to individual components
- expose wrappers for building new templates layer by layer
- improve typing so that all the above are easy to consume and compose new docs templates from
- fix useFetchDocs typing (properties array, rather than single property) and move component to more sensible location
- includes merged code from https://github.com/teambit/bit/pull/5790 (anything to do with composition-app - should be merged asap)
- Tested on a separate repo to make sure docs render correctly and can now be customised. Working well

### CODE REVIEW

File: scopes/react/ui/compositions-app/composition-app.docs.mdx [hunk 1]
```diff
@@ -0,0 +1,3 @@
+This mounts compositions into the DOM in the component preview.
```
Comment (GiladShoham) [inline scopes/react/ui/compositions-app/composition-app.docs.mdx:1]: why don't you have here the metadata section with labels and everything?

File: scopes/react/ui/docs-content/docs-content.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    <ErrorBoundary FallbackComponent={ErrorFallback}>
+    {Content.isMDXComponent ? (
+      <MDXLayout {...rest}>
+        <Content />
+      </MDXLayout>
+    ) : (
+      <Content />
+    )}
+  </ErrorBoundary>
+  );
```
Comment (KutnerUri) [inline scopes/react/ui/docs-content/docs-content.tsx:34]: prettify

File: scopes/react/ui/docs-content/docs-content.tsx [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+const defaultDocs = {
+  examples: [],
+  labels: [],
+  abstract: '',
+};
+
+export function DocsContent({ docs = defaultDocs, ...rest }: DocsContentProps) {
+
```
Comment (KutnerUri) [inline scopes/react/ui/docs-content/docs-content.tsx:21]:
```suggestion
```

File: scopes/react/ui/properties/properties.tsx [hunk 4]
```diff
@@ -1,9 +1,21 @@
 import { LinkedHeading } from '@teambit/documenter.ui.linked-heading';
 import { PropTable } from '@teambit/documenter.ui.property-table';
 import { Section } from '@teambit/documenter.ui.section';
+import { useFetchDocs } from '@teambit/component.ui.hooks.use-fetch-docs';
 import React from 'react';
 
-export function Properties({ properties }: any) {
+export type PropertiesParams = {
```
Comment (KutnerUri) [inline scopes/react/ui/properties/properties.tsx:7]: plz rename PropertiesParams -> PropertiesProps

File: scopes/react/ui/properties/properties.tsx [hunk 5]
```diff
@@ -1,9 +1,21 @@
...
+  componentId: string;
+}
+
+export function Properties({ componentId }: PropertiesParams) {
+  const { loading, error, data } = useFetchDocs(componentId);
```
Comment (KutnerUri) [inline scopes/react/ui/properties/properties.tsx:12]: moving the fetch here could potentially really improve loading times.

however, what if we want to use docs data in other places?  (ie in CompositionsSummery, etc)

File: .bitmap [hunk 6]
```diff
@@ -1500,6 +1518,12 @@
         "mainFile": "index.ts",
         "rootDir": "scopes/preview/ui/preview-placeholder"
     },
+    "ui/properties": {
```
Comment (KutnerUri) [inline .bitmap:1521]: maybe remame to properties table?

File: .bitmap [hunk 7]
```diff
@@ -1152,6 +1158,12 @@
         "mainFile": "index.ts",
         "rootDir": "scopes/compositions/ui/compositions-menu-bar"
     },
+    "ui/compositions-summary": {
```
Comment (KutnerUri) [inline .bitmap:1161]: I think you should rename to `composition-carousel` or `compositions-gallery`

File: .bitmap [hunk 8]
```diff
@@ -1020,6 +1020,12 @@
         "mainFile": "index.ts",
         "rootDir": "scopes/design/ui/alert-card"
     },
+    "ui/apply-providers": {
```
Comment (KutnerUri) [inline .bitmap:1023]: I think we should group these new components under a `ui/preview` or `ui/docs` namespace

**Review by KutnerUri — APPROVED** [review]

File: scopes/react/ui/apply-providers/apply-providers.tsx [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * base template for react component documentation
+ */
+export function ApplyProviders({ renderingContext, children, ...rest }: ApplyProvidersProps) {
+
+  const { providers = [] } = renderingContext.get(ReactAspect.id) || {};
+
+  return (
+    <div className={classNames(styles.docsMainBlock)} {...rest}>
```
[the remark this answers is not in the document]
Comment (benjgil) [reply scopes/react/ui/apply-providers/apply-providers.tsx:25]: done

File: scopes/react/ui/docs-app/docs-app.module.scss [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+}
+
+.compositionCard {
+  max-width: 100%;
+  overflow: hidden;
+  max-height: 500px; // limit the composition card height in the overview page
+}
+
+.mdx {
+  font-size: 16px;
```
[the remark this answers is not in the document]
Comment (benjgil) [reply scopes/react/ui/docs-app/docs-app.module.scss:12]: this has just been moved from where it was previously. Not in the scope of this PR to change/update

File: scopes/react/ui/docs-app/docs-app.tsx [hunk 11]
```diff
@@ -1,30 +1,54 @@
...
-  docs?: docsFile;
-  componentId: string;
-  compositions: [React.ComponentType];
-  renderingContext: RenderingContext;
+import { DocsTheme } from './docs-theme';
+import { ExamplesOverview } from './examples-overview';
+import type { DocsFile } from './examples-overview';
+import styles from './docs-app.module.scss';
+
+export type ReactDocsAppParams = {
```
[the remark this answers is not in the document]
Comment (benjgil) [reply scopes/react/ui/docs-app/docs-app.tsx:14]: interesting, was just following the syntax in these components :) 
happy to change it

File: scopes/react/ui/docs-app/docs-app.tsx [hunk 12]
```diff
@@ -1,30 +1,54 @@
...
+  componentId: string,
+  docs: DocsFile | undefined,
+  compositions: any,
```
[the remark this answers is not in the document]
Comment (benjgil) [reply scopes/react/ui/docs-app/docs-app.tsx:17]: this is just what was there. I dont want to get too deep into refactors if i'm not sure how it'll affect things downstream, not in the scope of this PR

File: scopes/react/ui/docs-app/docs-app.tsx [hunk 13]
```diff
@@ -1,30 +1,54 @@
...
+    componentId,
+    docs = defaultDocs,
+    compositions,
+    context
+  }: ReactDocsAppParams
+) {
+
+  // Next 2 lines are to support legacy code (ExamplesOverview)
+  const { examples = [] } = docs;
+  const Content: any = isFunction(docs.default) ? docs.default : () => null;
```
[the remark this answers is not in the document]
Comment (benjgil) [reply scopes/react/ui/docs-app/docs-app.tsx:37]: Again, just moving code around. I dont really want to hold up this big change to refactor all the code. This is to support a legacy feature that will likely be removed anyway, so not sure it's worth messing with here

File: scopes/react/ui/docs-app/docs.app-root.tsx [hunk 14]
```diff
@@ -5,32 +5,38 @@ import ReactDOM from 'react-dom';
...
-  /* docs: */ DocsFile | undefined,
-  /* compositions: */ Record<string, any>,
-  /* context: */ RenderingContext
-];
-
-export default function DocsRoot(
+/**
+ * These are the parameters supplied to all docs apps by the bit UI runtime
+ */
+export type ReactDocsRootType = (
```
[the remark this answers is not in the document]
Comment (benjgil) [reply scopes/react/ui/docs-app/docs.app-root.tsx:11]: I'll be refactoring this in a separate PR, where the params will be an object and then we can get rid of the function type, so i'll leave it to there

File: scopes/react/ui/docs-content/docs-content.tsx (hunk 2 above)
Comment (benjgil) [reply scopes/react/ui/docs-content/docs-content.tsx:33]: done (though it still looks pretty much the same - it's just how that syntax looks)

File: scopes/react/ui/properties/properties.tsx (hunk 4 above)
Comment (benjgil) [reply scopes/react/ui/properties/properties.tsx:7]: done

File: scopes/react/ui/properties/properties.tsx (hunk 5 above)
Comment (benjgil) [reply scopes/react/ui/properties/properties.tsx:12]: And i guess that's the point of the hook, you can use it wherever you need to (and it's memo-ised so shouldnt be a loading issue)

File: scopes/react/ui/compositions-app/composition-app.docs.mdx (hunk 1 above)
Comment (benjgil) [reply scopes/react/ui/compositions-app/composition-app.docs.mdx:1]: this is from Uri's changes - the rebase should remove this from the PR

File: scopes/react/ui/apply-providers/apply-providers.tsx [hunk 15]
```diff
@@ -1,28 +1,25 @@
...
- * base template for react component documentation
+ * applies providers from rendering context, and error boundary
  */
 export function ApplyProviders({ renderingContext, children, ...rest }: ApplyProvidersProps) {
 
   const { providers = [] } = renderingContext.get(ReactAspect.id) || {};
 
   return (
-    <div className={classNames(styles.docsMainBlock)} {...rest}>
+    <div {...rest}>
```
Comment (KutnerUri) [inline scopes/react/ui/apply-providers/apply-providers.tsx:22]: I think you can remove this div. If someone wants to add properties to ApplyProviders, he can just apply a div above it and get the same results.

File: scopes/react/ui/docs-app/docs-app.module.scss (hunk 10 above)
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/ui/docs-app/docs-app.module.scss:21]: sure.

**Review by KutnerUri — APPROVED** [review]

File: scopes/react/ui/properties/properties.tsx (hunk 5 above)
Comment (KutnerUri) [reply scopes/react/ui/properties/properties.tsx]: I guess. Good.

File: scopes/react/react/react.main.runtime.ts [hunk 16]
```diff
@@ -172,6 +172,7 @@ export class ReactMain {
    */
   overrideDocsTemplate(templatePath: string) {
     return this.envs.override({
+      getDevEnvId: (context: DevServerContext) => this.reactEnv.getDevEnvId(context.envDefinition.id),
```
[the remark this answers is not in the document]
Comment (KutnerUri) [reply scopes/react/react/react.main.runtime.ts:175]: ahh it's was a missing feature in react env. ok 💪

**Review by GiladShoham — APPROVED** [review]

**Pull request merged**